plugin-oar2/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+
+                                    self.control_handle.add_source(
+                                        self.plugin_name.clone(),
+                                        source_name,
+                                        new_source,
+                                        TriggerSpec::at_interval(self.poll_interval),
+                                    );
+                                }
+                            }
+                        }
+                    }
+                } else if let Err(e) = event {
+                    log::error!("watch error: {:?}", e);
+                }
+            }
+        }
+
+        let handler = JobDetector {
+            config_path: config_path.clone(),
+            cpu_metric,
+            memory_metric,
+            control_handle,
+            plugin_name,
+            poll_interval,
+        };
+        let mut watcher = notify::recommended_watcher(handler)?;
+
+        watcher.watch(&config_path.join("cpuacct/oar"), RecursiveMode::NonRecursive)?;
+        watcher.watch(&config_path.join("memory/oar"), RecursiveMode::NonRecursive)?;
+
+        self.watcher = Some(watcher);
+
+        Ok(())
+    }
+
+    fn stop(&mut self) -> anyhow::Result<()> {
+        Ok(())
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        let mut path = PathBuf::new();
+        path.push("/sys/fs/cgroup");
+        Self {
+            path,
+            poll_interval: Duration::from_secs(1),
+        }
+    }
+}
+
+impl Source for OarJobSource {
+    fn poll(&mut self, measurements: &mut MeasurementAccumulator, timestamp: Timestamp) -> Result<(), PollError> {
+        let cpu_usage_file = &mut self.cgroup_cpu_file;
+        cpu_usage_file.rewind()?;
+        let mut cpu_usage = String::new();
+        cpu_usage_file.read_to_string(&mut cpu_usage)?;
+        let memory_usage_file = &mut self.cgroup_memory_file;
+        memory_usage_file.rewind()?;
+        let mut memory_usage = String::new();
+        memory_usage_file.read_to_string(&mut memory_usage)?;
+        let cpu_usage_u64 = cpu_usage.trim().parse::<u64>()?;
+        let memory_usage_u64 = memory_usage.trim().parse::<u64>()?;
+
+        measurements.push(
+            MeasurementPoint::new(
+                timestamp,
+                self.cpu_metric,
+                Resource::LocalMachine,
+                ResourceConsumer::ControlGroup {
+                    path: (self.cpu_file_path.to_str().unwrap().to_owned().into()),
+                },
+                cpu_usage_u64,
+            )
+            .with_attr("oar_job_id", self.job_id),
+        );
+
+        measurements.push(
+            MeasurementPoint::new(
+                timestamp,
+                self.memory_metric,
+                Resource::LocalMachine,
+                ResourceConsumer::ControlGroup {
+                    path: (self.memory_file_path.to_str().unwrap().to_owned().into()),
```

Comment:
I think it's better to leave the path empty or log an error than panicking

```suggestion
                    path: (self.memory_file_path.to_str().unwrap_or_default().to_owned().into()),
```

Reply:
But `memory_file_path` has been obtained by the plugin before, it should in no case be invalid.

Reply:
Yes you're right, but what should we do ? Keep using unwrap because a panic would be a real panic if happening, or logging the error and skipping the wrong/weird value ?

Reply:
I'm tempted to keep unwrap (or use expect to provide an explanation of why it should never happen). Better detect the mistake early than propagating a weird value, isn't it?

Panics are caught by tokio anyway (i.e. it won't crash the whole pipeline).

Reply:
Yes, an expect would be good enough